Extension/gulpfile.js

```diff
@@ -290,6 +292,108 @@ gulp.task("translations-import", (done) => {
+                // Redirect - download from new location
+                let redirectUrl;
+                if (typeof response.headers.location === "string") {
+                    redirectUrl = response.headers.location;
+                } else {
+                    if (!response.headers.location) {
+                        console.log(`Invalid download location received`);
+                        return reject();
+                    }
+                    redirectUrl = response.headers.location[0];
+                }
+                console.log(`Using redirectUrl: '${redirectUrl}'`);
+                return resolve(DownloadFile(redirectUrl));
+            } else if (response.statusCode !== 200) {
+                if (response.statusCode === undefined || response.statusCode === null) {
+                    console.log("unknown error code.");
+                    return reject();
+                }
+                console.log(`failed with error code: '${response.statusCode}'`);
+                return reject();
+            }
+
+            response.on('data', (data) => {
+                buffers.push(data);
+            });
+
+            response.on('end', () => {
+                if (buffers.length > 0) {
+                    return resolve(Buffer.concat(buffers));
+                } else {
+                    return reject();
+                }
+            });
+
+            response.on('error', err => {
+                console.log(`problem with request: '${err.message}'`);
+                return reject();
+            });
+        });
+
+        req.on('error', err => {
+            console.log(`problem with request: '${err.message}'`);
+            return reject();
+        });
+
+        // Execute the request
+        req.end();
+    });
+    
+}
+
+async function generatePackageHashes(packageJson) {
+    const downloadAndGetHash = async (url) => {
+        console.log(url);
+        try {
+            const buf = await DownloadFile(url);
+            if (buf) {
+                const hash = crypto.createHash('sha256');
+                hash.update(buf);
+                const value = hash.digest('hex').toUpperCase();
+                return value;
+            }
+            return undefined;
+        } catch (err) {
+            return undefined;
+        }
+    };
+
+    for (let dependency of packageJson.runtimeDependencies) {
+        console.log(`-------- Downloading package: '${dependency.description}' --------`);
+        const hash = await downloadAndGetHash(dependency.url);
+        if (hash) {
+            dependency.integrity = hash;
+            console.log(`integrity: '${hash}'`);
+        } else {
+            console.log(`No hash generated for package '${dependency.description}`);
+        }
+        console.log(`\n`);
+    }
+
+    let content = JSON.stringify(packageJson, null, 2);
+    return content;
+}
+
+gulp.task('generate-package-hashes', async (done) => {
```

Comment:
I don't think anything else we do in gulpfile.js is 'async'.  Does this imply that when this script is run, it will return before the operation is completed?

Reply:
It would error without the async. I found that it needs there because there is an await.

Reply:
It looks like the await is required because DownloadFile was implemented using a Promise rather than sync.  Here, it would be fine if that were sync.  My concern is that being async, maybe the command being executed might not be finished before it returns, and the next command executed.  Maybe a race in which running generatePackageHashes returns before they are actually generated, and the next command (yarn run compile) picks up the file before it's ready.  Unlikely, but would be good to ensure there isn't a possible issue here.  The gulp process probably waits for 'done' to be signalled, I'm just not sure.

Reply:
Verified that the gulp task waits before running other tasks.